backend/src/crd/pkg/apis/viewer/v1alpha1/types.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Viewer is a specification for a Viewer resource.
+type Viewer struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec contains specifications that pertain to how the viewer is launched and
+	// managed by its controller.
+	Spec ViewerSpec `json:"spec"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ViewerList is a list of Viewer resources.
+type ViewerList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	// Items is a list of viewers.
+	Items []Viewer `json:"items"`
+}
+
+// ViewerType is the underlying type of the view. Currently, only Tensorboard is
+// explicitly supported by the Viewer CRD.
+type ViewerType string
+
+const (
+	// ViewerTypeTensorboard is the ViewerType constant used to indicate that the
+	// underlying type is Tensorboard. An instance named `instance123` will serve
+	// under /tensorboard/instance123.
```

Comment:
We should also consider mapping logdirs to instances in the route if we can, so that the user can just navigate to /tensorboard/encoded-paths-here, and they get routed to a new or existing instance by the CRD. Is this doable?

Reply:
I looked into this a bit, and at this point, I don't think we can. In that, the CRD isn't watching the routes, so it can't create a new one when the user navigates to it. The other thing is, I think it's worth having short names (right now, auto-generated by k8s) to give a shorter user-facing url. Mapping the logdirs to a name may result in a longer string, which I'm not in favour of.

Tangential to all this is keeping a unique a instance of TB for a given set of logdirs. I think this is a nice to have optimization, and I'm happy to look into adding this functionality in a follow up PR. What do you think?

Reply:
Who keeps track of that mapping though?

This isn't P0 by any means for this implementation, but it's good to keep it in mind while we're working on it.

Reply:
I imagine the right approach would be to label the viewer with the logdirs, which you can select for to ensure it's not already created. The reconciler here aims to be stateless as much as possible.

Reply:
SGTM.